execute/ring.go

```diff
@@ -0,0 +1,47 @@
+package execute
+
+type ring struct {
+	buf []interface{}
+	i   int
+	sz  int
+}
+
+func newRing(sz int) *ring {
+	return &ring{
+		buf: make([]interface{}, sz),
+	}
+}
+
+func (r *ring) Next() interface{} {
+	if r.sz == 0 {
+		return nil
+	}
+
+	v := r.buf[r.i]
+	r.buf[r.i] = nil
+	r.i++
+	if r.i >= len(r.buf) {
+		r.i = 0
+	}
+	r.sz--
+	return v
+}
+
+func (r *ring) Len() int {
+	return r.sz
+}
+
+func (r *ring) Append(v interface{}) {
+	if r.sz == len(r.buf) {
+		// Copy the existing buffer.
+		// We copy this in two parts.
+		buf := make([]interface{}, 0, r.sz*2)
+		buf = append(buf, r.buf[r.i:]...)
```

Comment:
not sure how go behaves here, but manually copying a buffer could slow down the performance? 
Kindly let me know if go behaves otherwise.

Reply:
It looks like I forgot to finish writing my comment. I don't believe manually copying a buffer slows anything down. These two statements are the same in Go.

```
newBuf := make([]int, len(buf))
copy(newBuf, buf)
```

```
newBuf := make([]int, 0, len(buf))
newBuf = append(newBuf, buf...)
```

Append's implementation is basically this:

```
func append(dst []int, item int) []int {
    if cap(dst) <= len(dst)+1 {
        dst = dst[:len(dst)+1]
    } else {
        cpy := make([]int, len(dst)+1, len(dst)*2)
        copy(cpy, dst)
        dst = cpy
    }
    dst[len(dst)-1] = item
}
```

It's a little more complex because it accepts variable arguments but it expands the capacity of a slice and then sets the value when there is capacity left and it copies the array into a new buffer when the append would exceed the capacity.

```
buf = append(buf, 1)

newBuf = make([]int, len(buf), len(buf)*2)
copy(,

Reply:
that's insightful. Thanks